api/src/services/application-flagged-set.service.ts

```diff
@@ -470,13 +468,229 @@ export class ApplicationFlaggedSetService implements OnModuleInit {
     return true;
   }
 
+  async processDuplicates(listingId?: string): Promise<SuccessDTO> {
+    this.logger.warn('running the Application flagged sets version 2 cron job');
+    await this.markCronJobAsStarted(CRON_JOB_NAME);
+    const duplicatesCloseDate =
+      !!process.env.DUPLICATES_CLOSE_DATE &&
+      dayjs(process.env.DUPLICATES_CLOSE_DATE, 'YYYY-MM-DD HH:mm Z');
+
+    if (!duplicatesCloseDate || duplicatesCloseDate > dayjs(new Date())) {
+      this.logger.warn(
+        'DUPLICATES_CLOSE_DATE either not set or is in the future',
+      );
+      return;
+    }
+    const outOfDateListings = await this.prisma.listings.findMany({
+      select: {
+        id: true,
+        afsLastRunAt: true,
+        name: true,
+      },
+      where: {
+        lastApplicationUpdateAt: {
+          not: null,
+        },
+        id: listingId,
+        AND: [
+          {
+            OR: [
+              // Only run this job on listings that were closed after the DUPLICATES_CLOSE_DATE
+              { closedAt: { gte: duplicatesCloseDate.toDate() } },
+              { closedAt: null },
+            ],
+          },
+          {
+            OR: [
+              {
+                afsLastRunAt: {
+                  equals: null,
+                },
+              },
+              {
+                afsLastRunAt: {
+                  lte: this.prisma.listings.fields.lastApplicationUpdateAt,
+                },
+              },
+            ],
+          },
+        ],
+      },
+    });
+    this.logger.warn(
+      `running duplicates check on ${outOfDateListings.length} listings`,
+    );
+    for (const listing of outOfDateListings) {
+      type PossibleFlaggedSetQuery = {
+        key: string;
+        type: RuleEnum;
+        applicationids: string[];
+      };
+      // find all flagged keys for this listing that applies to more than one listing
+      const flaggedApplicationGrouped: PossibleFlaggedSetQuery[] = await this
+        .prisma
+        .$queryRaw`select key, array_agg(application_id) as applicationIds, type
+        FROM application_flagged_set_possibilities
+        WHERE listing_id = ${listing.id}::UUID
+        GROUP BY key, type
+        HAVING count(application_id) > 1;`;
+
+      // group all of the groups by the applicationIds
+      const reduced = flaggedApplicationGrouped.reduce(
+        (accumulated, applicationGroup) => {
+          const key = applicationGroup.applicationids
+            .sort((a, b) => a.localeCompare(b))
+            .toString();
+          const values = accumulated[key] || [];
+          values.push(applicationGroup);
+          accumulated[key] = values;
+          return accumulated;
+        },
+        {} as {
+          [key: string]: PossibleFlaggedSetQuery[];
+        },
+      );
+
+      const applicationFlaggedSetsInDB =
+        await this.prisma.applicationFlaggedSet.findMany({
```

Comment:
I think on listings with large numbers of application flagged sets this may cause problems? 
I would suggest testing against prod data locally and keeping an eye on your activity monitor to see what happens to mem usage

Reply:
I didn't notice any spike in my memory when using the Fremont property. It does run in only a few seconds so potentially the spike is too short lived to notice. I think we'll want to monitor it though in heroku to better simulate the memory availability